### PR TITLE
Improve email address hostname validation

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb Ready for review.

File: core/lib/spree/core/validators/email.rb [hunk 1]
```diff
@@ -15,8 +15,10 @@ class EmailValidator < ActiveModel::EachValidator
       .to_set
     )
 
-  HOSTNAME_PATTERN =
-    /\A(?:[-\p{L}\d]{1,63}\.)+#{Regexp.union(*TLDS)}\z/i.freeze
+  edge  = /[[:alnum:]]/
+  label = /(?:#{edge}|#{edge}[[:alnum:]\-]{0,61}#{edge})/
+
+  HOSTNAME_PATTERN = /\A#{label}(?:\.#{label})*\.#{Regexp.union(*TLDS)}\z/.freeze
```
Comment (dkubb) [inline core/lib/spree/core/validators/email.rb]: For the `HOSTNAME_PATTERN` I was thinking perhaps something a bit stricter like:

```ruby
/\A(?=.{1,255}\z)((?=.{1,63}\.)[[:alnum:]](?:-?[[:alnum:]])*\.)+#{Regexp.union(*TLDS)}\z/
```

This has a few more desirable properties:

* The total hostname length is checked to make sure it's between 1 and 255 characters (the actual limits of the hostname according to RFC's -- I am aware that the email address limit is 254)
* The label length is checked to make sure it's between 1 and 63 characters
* An optional `-` is allowed to be between parts of a label, but never at the beginning or end.
* The dot must be at the end of each label, making the resulting regexp a bit simpler to read than the previous regexp with it's "edge" pieces.

Comment (mbj) [inline core/lib/spree/core/validators/email.rb]: I'll take that one, but make the repeated group a non capturing one.

Comment (dkubb) [inline core/lib/spree/core/validators/email.rb]: > I'll take that one, but make the repeated group a non capturing one.

Ah yeah, of course. That's an artifact of some local testing I was doing in irb.

**Comment by mbj** [comment]
@dkubb ready for the next pass.

**Pull request merged**